src/main/java/org/broadinstitute/hellbender/engine/AlignmentContext.java

```diff
@@ -97,45 +96,58 @@ public AlignmentContext stratify(final ReadOrientation type) {
+        return this.splitContextBySampleName((String) null, header);
     }
 
     /**
-     * Splits the given AlignmentContext into a StratifiedAlignmentContext per sample, but referencd by sample name instead
+     * Splits the given AlignmentContext into a StratifiedAlignmentContext per sample, but referenced by sample name instead
      * of sample object.
      *
      * @param assumedSingleSample If non-null, assume this is the only sample in our pileup and return immediately.
      *                            If null, get the list of samples from the provided header and do the work of splitting by sample.
-     * @return a Map of sample name to StratifiedAlignmentContext
-     *
+     * @return a Map of sample name to StratifiedAlignmentContext; samples without coverage are not included
      **/
     public Map<String, AlignmentContext> splitContextBySampleName(final String assumedSingleSample, final SAMFileHeader header) {
         if (assumedSingleSample != null){
             return Collections.singletonMap(assumedSingleSample, this);
         }
-
-        final Map<String, AlignmentContext> contexts = new HashMap<>();
-
-        for(final String sample: basePileup.getSamples(header)) {
-            final ReadPileup pileupForSample = basePileup.makeFilteredPileup(pe -> Objects.equals(sample, ReadUtils.getSampleName(pe.getRead(), header)));
-
-            if (sample == null){
-                throw new UserException.ReadMissingReadGroup(pileupForSample.iterator().next().getRead());
-            }
-
+        final Locatable loc = this.getLocation();
+        // this will throw an user error if there are samples without RG/sampleName
+        final Map<String, ReadPileup> pileups = this.getBasePileup().splitBySample(header, assumedSingleSample);
+        final Map<String, AlignmentContext> contexts = new HashMap<>(pileups.size());
+        for (final Map.Entry<String, ReadPileup> entry : pileups.entrySet()) {
             // Don't add empty pileups to the split context.
-            if(! pileupForSample.isEmpty()) {
-                contexts.put(sample, new AlignmentContext(loc, pileupForSample));
+            if (entry.getValue().isEmpty()) {
+                continue;
             }
+            contexts.put(entry.getKey(), new AlignmentContext(loc, entry.getValue()));
         }
-
         return contexts;
     }
 
     public static Map<String, AlignmentContext> splitContextBySampleName(final ReadPileup pileup, final SAMFileHeader header) {
         return new AlignmentContext(pileup.getLocation(), pileup).splitContextBySampleName(header);
     }
 
+    /**
+     * Splits the AlignmentContext into one context per read group
+     *
+     * @return a Map of ReadGroup to AlignmentContext, or an empty map if context has no base pileup
+     **/
+    public Map<SAMReadGroupRecord, AlignmentContext> splitContextByReadGroup(final Collection<SAMReadGroupRecord> readGroups) {
```

Comment:
This method was implemented in a previous commit (7f05fa095fd8e5cfe0aad4bfd2bc1945e8588780) by @droazen and removed by you in the improvement for a single sample (71c9cfdf841562a8f10fd844b4566579b4fb8db8). It is here because I've just miss this fact when merging branches. I'll remove it if it is not necessary.